### PR TITLE
chore(TS): Observable

### PR DESCRIPTION
**Pull request by ShaMan123**
- fixes observable off - now removes the handler instead of removing the handler after `fire`
- ~~`observable` tests pass accept 2 on active selection stuff~~ **FIXED `once` migration error**
- ~~I can't manage to extend Object/StaticCanvas with the class/proto/instance of class `Observable`. Nothing works.~~
- make it a class so it becomes a super class for Object/Canvas for overriding needs etc.

### CODE REVIEW

File: src/shapes/object.class.js [hunk 1]
```diff
@@ -1905,9 +1907,13 @@
...
 
-  fabric.util.createAccessors && fabric.util.createAccessors(fabric.Object);
+  // fabric.util.createAccessors && fabric.util.createAccessors(fabric.Object);
 
-  extend(fabric.Object.prototype, fabric.Observable);
+  fabric.Object.prototype.on = Observable.prototype.on;
+  fabric.Object.prototype.once = Observable.prototype.once;
+  fabric.Object.prototype.off = Observable.prototype.off;
+  fabric.Object.prototype.fire = Observable.prototype.fire;
+  fabric.Object.prototype._removeEventListener = Observable.prototype._removeEventListener;
```
Comment (asturur) [inline src/shapes/object.class.js:1916]: Why not Object.assign(fabric.Object.prototype, fabric.Observable) ?

File: src/util/lang_object.js [hunk 2]
```diff
@@ -72,5 +72,5 @@
     extend: extend,
     clone: clone
   };
-  fabric.util.object.extend(fabric.util, fabric.Observable);
+  // fabric.util.object.extend(fabric.util, fabric.Observable); //snot sure why this is here
```
Comment (asturur) [inline src/util/lang_object.js:75]: just trash it probably a relic of the past

File: src/mixins/observable.mixin.js [hunk 3]
```diff
@@ -1,145 +0,0 @@
...
-   * @namespace fabric.Observable
-   * @tutorial {@link http://fabricjs.com/fabric-intro-part-2#events}
-   * @see {@link http://fabricjs.com/events|Events demo}
-   */
-  fabric.Observable = {
-    fire: fire,
-    on: on,
-    once: once,
-    off: off,
-  };
```
Comment (asturur) [inline src/mixins/observable.mixin.js:144]: Observable was an object with 4 methods. Why does to become a class?

File: src/mixins/Observable.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
+//@ts-nocheck
+
+type EventRegistryObject = { [eventName: string]: Function };
```
Comment (asturur) [inline src/mixins/Observable.ts:3]: I think there is a a TS util, Record, that makes this more clear.

File: test/unit/observable.js [hunk 5]
```diff
@@ -1,19 +1,19 @@
 QUnit.module('fabric.Observable');
 
-QUnit.test('fabric.Observable exists', function(assert) {
+QUnit.test('fabric.Observable exists', function (assert) {
   assert.ok(fabric.Observable);
-  assert.ok(fabric.Observable.fire);
-  assert.ok(fabric.Observable.on);
-  assert.ok(fabric.Observable.once);
-  assert.ok(fabric.Observable.off);
+  var o = new fabric.Observable();
```
Comment (asturur) [inline test/unit/observable.js:5]: i undertstand this looks better, but is a mixin really supposed to to be initialized on its own?

File: index.js [hunk 6]
```diff
@@ -1,100 +1,105 @@
...
-import './src/mixins/itext.svg_export.js'; // optional itext
-import './src/shapes/textbox.class.js'; // optional textbox
-import './src/mixins/default_controls.js'; // optional interaction
+import { fabric } from './HEADER';
+// import './lib/event'), // optional gestures
+import './src/mixins/collection.mixin';
+
+// import './src/mixins/Observable';
+import { Observable } from './src/mixins/Observable';
+fabric.Observable = Observable;
```
Comment (asturur) [inline index.js:7]: i think is safe to bring all of those up, Observable does not seem to need to come after collection.mixin.
I would also rather do the assign to fabric inside the file itself.

Comment (ShaMan123) [reply index.js:7]: > I would also rather do the assign to fabric inside the file itself.

I am against for a few reasons:
- after we complete the migration we won't be assigning to fabric
- limiting the mess to the index which will change a lot anyways

File: src/shapes/object.class.js (hunk 1 above)
Comment (ShaMan123) [reply src/shapes/object.class.js:1916]: I tried that but it doesn't work

File: src/mixins/observable.mixin.js (hunk 3 above)
Comment (ShaMan123) [reply src/mixins/observable.mixin.js:144]: my intention is to make CommonMethods extend Observable

File: src/mixins/Observable.ts (hunk 4 above)
Comment (ShaMan123) [reply src/mixins/Observable.ts:3]: great

File: test/unit/observable.js (hunk 5 above)
Comment (ShaMan123) [reply test/unit/observable.js:5]: > my intention is to make CommonMethods extend Observable

File: src/shapes/object.class.js (hunk 1 above)
Comment (ShaMan123) [reply src/shapes/object.class.js]: because methods aren't enumerable on class proto

File: src/mixins/observable.mixin.ts [hunk 7]
```diff
@@ -1,146 +1,157 @@
...
+      const eventName = arg0;
       if (!this.__eventListeners[eventName]) {
         this.__eventListeners[eventName] = [];
       }
       this.__eventListeners[eventName].push(handler);
+      return () => this.off(eventName, handler);
+    }
+    else {
+      // noop
+      return () => false;
```
Comment (ShaMan123) [inline src/mixins/observable.mixin.ts:47]: not sure we need this
just in case a dev does something illegal

**Review by ShaMan123 — COMMENTED** [review]
READY

File: test/unit/observable.js (hunk 5 above)
Comment (ShaMan123) [reply test/unit/observable.js:5]: line node's `EventEmitter`

**Comment by ShaMan123** [comment]
closed by #8322

**Pull request closed without merging**